### PR TITLE
Do not accept interpolated types as traits in trait impls

### PR DESCRIPTION
**Pull request by petrochenkov**
Fixes https://github.com/rust-lang/rust/issues/46438

Needs crater run before proceeding.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
cc @rust-lang/release -- needs crater

**Comment by Mark-Simulacrum** [comment]
Crater run started.

**Comment by Mark-Simulacrum** [comment]
Crater results: http://cargobomb-reports.s3.amazonaws.com/pr-48052/index.html. 'Blacklisted' crates (spurious failures etc) can be found
[here](https://github.com/rust-lang-nursery/crater/blob/master/blacklist.md).
If you see any spurious failures not on the list, please make a PR against that file.

**Comment by petrochenkov** [comment]
6 legit regressions:
- downcast-rs-1.0.0
- faster-0.4.3
- glm-0.2.3
- glm_color-0.1.2
- guilt-by-association-0.4.1
- sql-0.4.2

I think we can deprecate this.
I'll change the error into a warning (no lints in the parser unfortunately), add a help message suggesting to replace `ty` with `path` and send PRs to affected crates.

**Comment by Mark-Simulacrum** [comment]
Reverse deps for breakage: https://gist.github.com/Mark-Simulacrum/e06765d8cd4a324da9fd47e4acf3fce7. I agree that it seems reasonable to deprecate this -- very few crates in the ecosystem are broken by this.

**Comment by nikomatsakis** [comment]
Hmm, I'm of two minds here. To be clear, we are specifically talking about deprecating **this** behavior, right?

```rust
macro_rules! foo {
  ($t:ty) => {
    impl $t for ... { }
  }
}
```

This would be deprecated because `$t`  is a "type", not a trait. I agree with that distinction and I do think this is roughly consistent with how we do things in macros. I just don't like how macros do things.

In particular, I don't like that it *matters* what kind of fragment you use -- and hence that `$t` is (internally) replaced with a "special type token" that can only appear in type position and so forth. I find this kind of makes macros a pain in the neck and breaks my mental model pretty regularly (which is basically one of token tree interpolation).

OTOH, I'm not sure if we have any room to change this until a hypothetical Macros 2.0, and I do feel like -- given that we have type tokens -- it is sort of "wrong" to use them in trait position. And not that many crates are affected.

OTOOH, it *does work* today, and it seems like the kind of kludge we could keep working. @petrochenkov can you elaborate a bit more on the motivation here?

**Comment by petrochenkov** [comment]
This PR is updated, PRs to affected crates are sent.

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -5547,8 +5548,16 @@ impl<'a> Parser<'a> {
 
                 let ty_first = ty_first.into_inner();
                 let path = match ty_first.node {
-                    // This notably includes paths passed through `ty` macro fragments (#46438).
-                    TyKind::Path(None, path) => path,
+                    TyKind::Path(None, path) => {
+                        if let Some(span) = ty_first_interpolated {
+                            self.diagnostic()
+                                .struct_span_warn(span, "expected a trait, found type")
+                                .note("this warning will become a hard error in a future release")
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: Is there a reason we can't use the standard lint infrastructure?

Comment (petrochenkov) [reply src/libsyntax/parse/parser.rs]: It's too early in the parser, we don't even have `NodeId`s yet to attach the lint.

**Comment by sgrif** [comment]
Just out of curiosity, why is `path` considered more appropriate here than `ty`? `path` also allows all sorts of things that aren't traits (I guess `path` is a strict subset of `ty`, but still...)

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/libsyntax/parse/parser.rs]: OK. Before we merge, let's try to [follow the other steps though](https://forge.rust-lang.org/rustc-bug-fix-procedure.html), i.e., create a tracking issue and so forth.

**Comment by nikomatsakis** [comment]
@sgrif 

> Just out of curiosity, why is path considered more appropriate here than ty? path also allows all sorts of things that aren't traits (I guess path is a strict subset of ty, but still...)

A fair question. Using "type" here feels sort of like a "category error" -- that is, the idea is roughly that when you parse a trait as a type, you convert it into `dyn Trait`, so this is "as wrong" as writing `impl dyn Debug for Foo`.

(That said, I still feel like macros *should* behave as if they were more-or-less copying tokens to the extent possible, in which case the actual fragment you use wouldn't matter.)

**Comment by nikomatsakis** [comment]
Nominating for discussing in @rust-lang/lang meeting. 

[To clean up, or not to clean up?](https://github.com/rust-lang/rust/pull/48502#issuecomment-371541882)

**Comment by eddyb** [comment]
I prefer the fragment distinction to be only used for matching and erased when expanding.
cc @jseyfried

**Comment by petrochenkov** [comment]
That would be a pretty serious shift from the current scheme though.
(You'd have to either re-tokenize everything parsed by macro matchers like `ty` back from AST, or somehow throw away the AST produced during parsing/matching and reuse the orinal tokens.)
(Anyway, this PR operates under assumptions of the current rules.)

**Comment by eddyb** [comment]
@petrochenkov We already keep the original tokens for at least items, I believe.
Although I supposed we haven't fully moved in the direction of caching parse results.

**Comment by nikomatsakis** [comment]
We discussed this in the @rust-lang/lang meeting today. I believe the consensus (correct me if I am mistaken) was that it generally makes sense to move towards a future where the precise fragment specifier that you used doesn't matter so much: so e.g. writing `$t:ty` doesn't mean you can't use those same tokens later as something else. We probably can't actually make macro rules behave that way entirely -- that would have to wait until Macros 2.0 -- but we don't have to close off cases where it *does* work today.

(Note though: we should not accept `dyn Trait` in these positions. =)

(As a side argument, it strikes me that -- for the same reason that we accept this in the parser -- it's perhaps convenient for macro authors to be able to use `ty` as a "cover grammar" for types and traits.)

Moreover, on the other side of the scale, it doesn't seem like this particularly simplifies implementation (more the opposite: we need an extra check), nor enables any future uses.

Therefore I move to close. (But feel free @petrochenkov or others to raise counterarguments.)

@rfcbot fcp close

**Comment by nikomatsakis** [comment]
ping @rust-lang/lang -- FCP ticky marks await you!

**Comment by scottmcm** [comment]
Given the crazy things that are possible with macro_rules, I have no problem with close here.

That said, I'd love some kind of a "what do capture kinds even mean" intent document for macros 2.0, because it still feels really weird to me that you can use a capture as something totally different from what you said it was in the capture kind.  (If it was just token based that would make sense, but iirc it's not in general...)

**Comment by nikomatsakis** [comment]
@scottmcm 

> That said, I'd love some kind of a "what do capture kinds even mean" intent document for macros 2.0

Agreed -- macros 2.0 definitely needs more definition, it's in a bit of "see what you want to see" situation.

> (If it was just token based that would make sense, but iirc it's not in general...)

As I said in the meeting, I definitely feel like we should try to ensure you can think of them as token trees to the extent possible (modulo expression precedence, as discussed in ... some RFC or other). That is, `$lhs * $rhs` should be equivalent to `($lhs) * ($rhs)`, even when e.g. lhs = `x + y`. =)

**Comment by eddyb** [comment]
@scottmcm I think that's consistent, in that you don't have *tokens* but one token ("tree") grouping together several tokens. That is, you can't take apart a fragment or "insert" anything into it.

An interesting example that I would *hope* to work in the future, is taking `(A, B)` as an expression, and putting it into a type position, having it reparsed as a type.
OTOH, I don't want a captured `X<Y>` (as a type) allowed in expression position followed by another expression, because that would reinterpret the second `>`.

I'm not sure how to formalize this, but one way to put it is that "syntactical elements used in the same level by the AST cannot come from different 'captured bags' of tokens".

This would imply that e.g. an expression followed by a "captured bag" of more than one token can *never* parse (although maybe some people would want `foo $tuple` to be a function call?).

**Pull request closed without merging**